bangc-ops/mlu_op.h

```diff
@@ -5821,6 +5821,107 @@ mluOpFill_v3(mluOpHandle_t handle,
+ * returns the results in the output tensor \b grad_input .
+ *
+ * @param[in] handle
+ * Handle to an MLUOP context that is used to manage MLU devices and queues in
+ * ::mluOpMoeDispatchBackwardData. For detailed information, see ::mluOpHandle_t.
+ * @param[in] gates_desc
+ * The descriptor of \b gates tensor, which contains dimension, data type and data layout.
+ * For detailed information, see ::mluOpTensorDescriptor_t.
+ * @param[in] gates
+ * Pointer to the MLU memory that stores the \b gates tensor.
+ * @param[in] indices_desc
+ * The descriptor of \b indices tensor, which contains dimension, data type and data layout.
+ * For detailed information, see ::mluOpTensorDescriptor_t.
+ * @param[in] indices
+ * Pointer to the MLU memory that stores the \b indices tensor.
+ * @param[in] locations_desc
+ * The descriptor of \b locations tensor, which contains dimension, data type and data layout.
+ * For detailed information, see ::mluOpTensorDescriptor_t.
+ * @param[in] locations
+ * Pointer to the MLU memory that stores the \b locations tensor.
+ * @param[in] dispatch_desc
+ * The descriptor of \b dispatch tensor, which contains dimension, data type and data layout.
+ * For detailed information, see ::mluOpTensorDescriptor_t.
+ * @param[in] dispatch
+ * Pointer to the MLU memory that stores the \b dispatch tensor.
+ * @param[in] samples
+ * The number of elements in the \b gates tensor, the \b indices tensor and the \b locations tensor.
+ * @param[in] capacity
+ * The maximum number of inputs that experts can process.
+ * @param[in] hidden
+ * The vector length of a single token.
+ * @param[in] num_experts
+ * The number of experts.
+ * @param[in] grad_input_desc
+ * The descriptor of \b grad_input tensor, which contains dimension, data type and data layout.
+ * For detailed information, see ::mluOpTensorDescriptor_t.
+ * @param[out] grad_input
+ * Pointer to the MLU memory that stores the \b grad_input tensor.
+ *
+ * @par Return
+ * - ::MLUOP_STATUS_SUCCESS, ::MLUOP_STATUS_BAD_PARAM, ::MLUOP_STATUS_ARCH_MISMATCH, ::MLUOP_STATUS_NOT_SUPPORTED
+ *
+ * @par Data Type
+ * - This function supports the following data types for input tensor \b gates , \b indices ,
+ *   \b locations , \b dispatch and output tensor \b grad_input .
+ *   - gates tensor: half, float
+ *   - indices tensor: int32
+ *   - locations tensor: int32
+ *   - dispatch tensor: half, float
+ *   - grad_input tensor: half, float
+ *
+ * @par Data Layout
+ * - The supported layout of the input tensors and output tensors must be \p MLUOP_LAYOUT_ARRAY.
+ *
+ * @par Scale Limitation
+ * - The first dimension of \b gates tensor, \b indices tensor, \b locations tensor and \b grad_input
+ *   tensor must be the same size and equal to \b samples .
+ * - The second dimension of \b grad_input tensor and \b dispatch tensor must be equal to \b hidden .
+ * - The first dimension of \b dispatch tensor must be equal to the multiplication result of
+ *   the \b capacity and \b num_experts .
+ *
```

Comment:
* - The value of the input parameters \b samples, \b capacity, \b hidden and \b num_experts
 *   must be greater than or equal to 0.
 * - The value range of the input parameter \b indices tensor must be greater than or equal to 0 and less than
 *   \b num_experts.
 * - The value range of the input parameter \b locations tensor must be greater than or equal to 0 and less than
 *   \b capacity.